fix #9405 by adding api design docs (#9455)
aspnet/Docs: aspnetcore/signalr/api-design.md

When the old client sends a single parameter, the extra `Param2` property will be left `null`. You can detect a message sent by an older client by checking the `Param2` for `null` and apply a default value. A new client can send both parameters.

[!code-typescript[CallWithObjectNew](api-design/sample/Samples.ts?name=CallWithObjectNew)]

The same technique works for methods defined on the client. You can send a custom object from the server side:

[!code-csharp[ClientSideObjectBasedOld](api-design/sample/Samples.cs?name=ClientSideObjectBasedOld)]

On the client side, you access the `Message` property rather than using a parameter:

[!code-typescript[OnWithObjectOld](api-design/sample/Samples.ts?name=OnWithObjectOld)]

If you later decide to add the sender of the message to the payload, add a property to the object:

[!code-csharp[ClientSideObjectBasedNew](api-design/sample/Samples.cs?name=ClientSideObjectBasedNew&highlight=5)]

The older clients won't be expecting the `Sender` value, so they'll ignore it. A new client can accept it by updating to read the new property:

[!code-typescript[OnWithObjectNew](api-design/sample/Samples.ts?name=OnWithObjectNew&highlight=1-4)]

> In this case, the new client is also tolerant of an old server that doesn't provide the `Sender` value, so it checks to see if it exists before accessing it.

Reviewer: - Split this into 2 sentences.
- In the 2nd part of the sentence, clarify what "it" is.

---
title: SignalR API design considerations
author: anurse

> description: 

Reviewer: Need a brief description here explaining what the doc focuses on.

monikerRange: '>= aspnetcore-2.1'
ms.author: anurse
ms.custom: mvc
ms.date: 11/05/2018
uid: signalr/api-design
---

# SignalR API design considerations

By [Andrew Stanton-Nurse](https://twitter.com/anurse)

This article provides guidance for building SignalR-based APIs.

## Use custom object parameters to ensure backwards-compatibility

Adding parameters to a SignalR hub method (on either the client or the server) is a *breaking change*, which means older clients/servers will get errors when they try to invoke the method without the appropriate number of parameters. However, adding properties to a custom object parameter is **not** a breaking change. This can be used to design compatible APIs that are resiliant to changes on the client or the server.

For example, consider a server-side API like the following:

[!code-csharp[ParameterBasedOldVersion](api-design/Samples.cs?name=ParameterBasedOldVersion)]

> The JavaScript client calls this method using `invoke`, like this:

Reviewer: ```suggestion
The JavaScript client calls this method using `invoke` as follows:
```

[!code-typescript[CallWithOneParameter](api-design/Samples.ts?name=CallWithOneParameter)]

If you later add a second parameter to the server method, older clients won't provide this parameter value. For example:

[!code-csharp[ParameterBasedNewVersion](api-design/Samples.cs?name=ParameterBasedNewVersion)]

When the old client tries to invoke this method, it will get an error like this:

```
Microsoft.AspNetCore.SignalR.HubException: Failed to invoke 'GetTotalLength' due to an error on the server.
```

On the server, you'll see a log message like this:

```
System.IO.InvalidDataException: Invocation provides 1 argument(s) but target expects 2.
```

The old client only sent one parameter, but the newer server API required two parameters. Using custom objects as parameters gives you more flexibility. Let's redesign the original API to use a custom object:

[!code-csharp[ObjectBasedOldVersion](api-design/Samples.cs?name=ObjectBasedOldVersion)]

Now, the client uses an object to call the method:

[!code-typescript[CallWithObject](api-design/Samples.ts?name=CallWithObject)]

> Now, instead of adding a parameter, we can add a property to the `GetTotalLengthRequest` object:

Reviewer: ```suggestion
Instead of adding a parameter, add a property to the `GetTotalLengthRequest` object:
```

> Instead of adding a parameter, add a property to the `GetTotalLengthRequest` object:

Reviewer: `GetTotalLengthRequest` sounds like a method name. Consider renaming the class to `TotalLengthRequest`.

[!code-csharp[ObjectBasedNewVersion](api-design/Samples.cs?name=ObjectBasedNewVersion&highlight=31,36-41)]

> When the old client sends us a single parameter, the extra `Param2` property will be left `null` and we can handle that in our code. A new client can send us both parameters.

Reviewer: Clarify what's meant by "we can handle that in our code". Handle what?

Author: Good point, pushed a new iteration with a bit more clarity.

[!code-typescript[CallWithObjectNew](api-design/Samples.ts?name=CallWithObjectNew)]

The same technique works for methods defined on the client. You can send a custom object from the server side:

[!code-csharp[ClientSideObjectBasedOld](api-design/Samples.cs?name=ClientSideObjectBasedOld)]

On the client side, you access the `Message` property rather than using a parameter:

[!code-typescript[OnWithObjectOld](api-design/Samples.ts?name=OnWithObjectOld)]

> If we later decide to add the sender of the message to the payload we send, we can do that by adding a property to the object:

Reviewer: ```suggestion
If you later decide to add the sender of the message to the payload, add a property to the object:
```

[!code-csharp[ClientSideObjectBasedNew](api-design/Samples.cs?name=ClientSideObjectBasedNew&highlight=60)]

> The older clients won't be expecting the `Sender` value, but they'll just ignore it. A new client can accept it by updating to read the new property:

Reviewer: ```suggestion
The older clients won't be expecting the `Sender` value, so they'll ignore it. A new client can accept it by updating to read the new property:
```

> [!code-typescript[OnWithObjectNew](api-design/Samples.ts?name=OnWithObjectNew&highlight=21-24)]

Reviewer 2: I don't see the highlight for any of your snippets

Author: Me neither ☹️

Author: Maybe `highlight` isn't supported with `name`? I think I specified everything right: https://dotnet.github.io/docfx/spec/docfx_flavored_markdown.html?tabs=tabid-1%2Ctabid-a#code-snippet

Author: @scottaddie do you know what might be going on here?

Reviewer: Not seeing anything obviously wrong here, but I'll continue to dig. On a related note, can you please move the sample code files into a signalr/api-design/sample folder? That's the convention we've followed elsewhere.

Reviewer: I see the problem now. Numbering resets within each region. In the case of the OnWithObjectNew region, your highlighting numbers should be 2-5.

> In this case, the new client is also tolerant of connecting to an old client that doesn't provide the `Sender` value, so it checks to see if it exists before accessing it.

Reviewer 2: "connecting" -> "talking" or "communicating"?

Author: Also it's not tolerant of an old client :). It's tolerant of an old server.

---

> title: SignalR API Design Considerations

Reviewer: ```suggestion
title: SignalR API design considerations
```

author: anurse
description: 
monikerRange: '>= aspnetcore-2.1'
ms.author: anurse
ms.custom: mvc
ms.date: 11/05/2018
uid: signalr/api-design
---

> # SignalR API Design Considerations

Reviewer: ```suggestion
# SignalR API design considerations
```

By [Andrew Stanton-Nurse](https://twitter.com/anurse)

> This article provides guidance for building SignalR-based APIs

Reviewer: End the sentence with a period.

## Use custom object parameters to ensure backwards-compatibility

Adding parameters to a SignalR hub method (on either the client or the server) is a *breaking change*, which means older clients/servers will get errors when they try to invoke the method without the appropriate number of parameters. However, adding properties to a custom object parameter is **not** a breaking change. This can be used to design compatible APIs that are resiliant to changes on the client or the server.

For example, consider a server-side API like the following:

[!code-csharp[ParameterBasedOldVersion](api-design/Samples.cs?name=ParameterBasedOldVersion)]

The JavaScript client calls this method using `invoke`, like this:

[!code-typescript[CallWithOneParameter](api-design/Samples.ts?name=CallWithOneParameter)]

> If you later add a second parameter to the server method, older clients will not provide this parameter value. For example:

Reviewer: will not --> won't

[!code-csharp[ParameterBasedNewVersion](api-design/Samples.cs?name=ParameterBasedNewVersion)]

When the old client tries to invoke this method, it will get an error like this:

```
Microsoft.AspNetCore.SignalR.HubException: Failed to invoke 'GetTotalLength' due to an error on the server.
```

> On the server, you will see a log message like this:

Reviewer: you will --> you'll

[!code-typescript[CallWithObjectNew](api-design/Samples.ts?name=CallWithObjectNew)]

The same technique works for methods defined on the client. You can send a custom object from the server side:

[!code-csharp[ClientSideObjectBasedOld](api-design/Samples.cs?name=ClientSideObjectBasedOld)]

On the client side, you access the `Message` property rather than using a parameter:

[!code-typescript[OnWithObjectOld](api-design/Samples.ts?name=OnWithObjectOld)]

If we later decide to add the sender of the message to the payload we send, we can do that by adding a property to the object:

[!code-csharp[ClientSideObjectBasedNew](api-design/Samples.cs?name=ClientSideObjectBasedNew&highlight=60)]

The older clients won't be expecting the `Sender` value, but they'll just ignore it. A new client can accept it by updating to read the new property:

[!code-typescript[OnWithObjectNew](api-design/Samples.ts?name=OnWithObjectNew&highlight=21-24)]

In this case, the new client is also tolerant of connecting to an old client that doesn't provide the `Sender` value, so it checks to see if it exists before accessing it.

> ## Provide detailed error information using `HubException`

Author: Yeah, as per in-person discussions I think I'll split it off as well.
